### PR TITLE
[MU4] Implement automatic theme switching for Windows

### PR DESCRIPTION
**Pull request by cbjeukendrup**
Implement automatic theme switching for Windows:

Windows automatic theme switching
(Just like on macOS, MuseScore switches much more fluent to dark theme than the OS itself. That must be an advantage of QML 😊)

Note: The window title bar doesn't yet switch. That would be a nice next step for the future, but for now, this was really enough Windows for me.

It was the first time that I wrote code with multiple threads. I hope I didn't do dangerous things, and would really appreciate a review :)

Furthermore, I don't have a computer available with Windows 7, so if somebody would want to test if this PR builds and runs correctly on Windows 7 too, that would be great. 

If this PR gets merged earlier than PR #7447, that PR has to be updated, or the other way round. 

Follow-up for PR #7366 (kind of).

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/CodeGuidelines.md)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [ ] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: src/framework/ui/internal/platform/windows/windowsplatformtheme.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    bool isDarkMode() const override;
+    async::Channel<bool> darkModeSwitched() const override;
+
+    void setAppThemeDark(bool) override;
+
+private:
+    async::Channel<bool> m_darkModeSwitched;
+    int m_buildNumber = 0;
+    bool isListening = false;
```
Comment (igorkorsukov) [inline src/framework/ui/internal/platform/windows/windowsplatformtheme.h:50]: needs rename to `m_isListening` and better use `std::atomic<bool>` in this case

File: src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                     KEY_NOTIFY | KEY_CREATE_SUB_KEY | KEY_ENUMERATE_SUB_KEYS | KEY_QUERY_VALUE | KEY_WOW64_64KEY,
+                     &hKey) == ERROR_SUCCESS) {
+        std::thread listenThread([this]() { listen(); });
+        listenThread.detach();
+    } else {
+        LOGD() << "Failed opening key for listening dark theme changes.";
+    }
+}
+
+void WindowsPlatformTheme::listen()
```
Comment (igorkorsukov) [inline src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp:85]: better rename to `th_listen` (this is not rule, but this helpful for easy understand)

File: src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+WindowsPlatformTheme::~WindowsPlatformTheme()
+{
+    stopListening();
+}
+
+void WindowsPlatformTheme::init()
+{
+    updateListeningStatus(configuration()->preferredThemeType());
+
+    configuration()->preferredThemeTypeChanged().onReceive(nullptr, [this](const ThemeType& newThemeType) {
```
Comment (igorkorsukov) [inline src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp:55]: better add in the `windowsplatformtheme.h `

```
#include "async/asyncable.h"

class WindowsPlatformTheme : public IPlatformTheme, public async::Asyncable 
```

and here 

```
configuration()->preferredThemeTypeChanged().onReceive(this, [this](const ThemeType& newThemeType) {
```

it is an automatic disconnect mechanism

**Comment by cbjeukendrup** [comment]
Thank you @igorkorsukov ! I updated the PR with your feedback.

**Review by RomanPudashkin — APPROVED** [review]

File: src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+{
+    if (m_isListening) {
+        return;
+    }
+    m_isListening = true;
+    if (RegOpenKeyEx(HKEY_CURRENT_USER, windowsThemeKey.c_str(), 0,
+                     KEY_NOTIFY | KEY_CREATE_SUB_KEY | KEY_ENUMERATE_SUB_KEYS | KEY_QUERY_VALUE | KEY_WOW64_64KEY,
+                     &hKey) == ERROR_SUCCESS) {
+        std::thread listenThread([this]() { th_listen(); });
+        listenThread.detach();
```
Comment (igorkorsukov) [inline src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp:79]: This moment can lead to a potential crash - if, at the time of closing the application when the WindowsPlatformTheme object is destroyed, a call to its class member (`m_darkModeSwitched`). In this case, the probability of failure is extremely small.
But in general, it is better to do this in such cases:
1. Make the thread a member of the class
2. In the `stopListening` method, do this

```
void WindowsPlatformTheme::stopListening()
{
    m_isListening = false;
    m_listenThread.join();
}
```

Comment (cbjeukendrup) [reply src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp:79]: But if I would put m_listenThread.join() in stopListening, then it blocks the app each time when you disable automatic theme switching, for at most 4000 ms (or whatever time we set in WaitForSingleObject). Maybe another option would be to call it only in the destructor, but I think that could theoretically still cause problems.

But, I just found a potentially very useful StackOverflow post (https://stackoverflow.com/a/6033171/11432719), so I will try that.

Comment (cbjeukendrup) [reply src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp]: My plan worked: now a "Stop" event is sent, so that `WaitForMultipleObjects` will return directly when it receives the event and the thread finishes (almost) directly, so calling `join()` won't cause waiting any more. Now it should be a bit safer!

Comment (igorkorsukov) [reply src/framework/ui/internal/platform/windows/windowsplatformtheme.cpp]: Well done!

**Review by igorkorsukov — APPROVED** [review]
